File: Introduction_to_OpenMP/Solutions/ftn/src/pi-serial.f90
! Calculating Pi with using method: 
!     integral 4.0/(1+x^2) dx = p	     
! 	   Serial Version	     
program pi_serial
implicit none
integer(kind=8) :: num_steps = 100000000000
integer(kind=8) i
real*8 :: PI_ref=3.1415926535897932,step,x,pi
real*8 :: summation=0.0
real   :: start,finish
call cpu_time(start)
step = 1/(1.0 * num_steps)

	do i =1,num_steps
	x=((i-1)+0.5) * step
summation=summation + (4.0/(1.0+(x*x)))
	end do
	pi = step * summation
call cpu_time(finish)
	write(*,*) "Estimated value of pi =",pi
	write(*,*) "Error =", abs(PI_ref - pi)
	write(*,*) "Compute time=", (finish-start)," seconds"
	end

File: Introduction_to_OpenMP/Solutions/c/src/pi-serial.c
/********************************************************
 * Calculating Pi with using midpoint rectangle method: *
 *  integral 4.0/(1+x^2) dx = pi	     		*
 *  or 							*
 *  summation (for i=0 to n) 4.0/(1+x_i^2)              *
 *  	where x_i = (i + 1/2) * delta x			*
 * 	   Serial Version	     			*
 ********************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <omp.h>

// Reference value to estimate error
double PI_ref=3.1415926535897932;

static long long num_rects=100000000000;
double step;
int main (int argc, char* argv[]){
  double start,stop;
  double diff_time;

  long long i;
  double x,pi,sum=0.0;

  start= omp_get_wtime();   
  step = 1.0 / (double) num_rects;
        
  for (i=0; i<num_rects; i++){
        x=(i+0.5) * step;
        sum = sum + 4.0/(1.0+(x*x));
  }

  pi = step * sum ;
  stop=omp_get_wtime();
  diff_time = stop-start;
  printf ("Estimated value of pi = %2.16lf\n",pi);
  printf ("Error = %2.16lf\n",fabs(PI_ref - pi));
  printf ("Compute time= %2.5lf seconds\n",diff_time);

}
